renderdoc/driver/vulkan/wrappers/vk_cmd_funcs.cpp

```diff
@@ -1493,28 +1493,48 @@ void WrappedVulkan::vkCmdBeginRenderPass(VkCommandBuffer commandBuffer,
     record->MarkResourceFrameReferenced(GetResID(pRenderPassBegin->renderPass), eFrameRef_Read);
 
     VkResourceRecord *fb = GetRecord(pRenderPassBegin->framebuffer);
+    VkResourceRecord *rp = GetRecord(pRenderPassBegin->renderPass);
 
     record->MarkResourceFrameReferenced(fb->GetResourceID(), eFrameRef_Read);
 
     rdcarray<VkImageMemoryBarrier> &barriers = record->cmdInfo->rpbarriers;
 
     barriers.clear();
 
-    if(fb->imageAttachments[0].barrier.sType && fb->imageAttachments[0].record)
+    FramebufferInfo *fbInfo = fb->framebufferInfo;
+    RenderPassInfo *rpInfo = rp->renderPassInfo;
+
+    if(fbInfo->imageAttachments[0].barrier.sType && fbInfo->imageAttachments[0].record)
     {
-      for(size_t i = 0; fb->imageAttachments[i].barrier.sType; i++)
+      bool renderArea_covers_entire_framebuffer =
+          pRenderPassBegin->renderArea.offset.x == 0 && pRenderPassBegin->renderArea.offset.y == 0 &&
+          pRenderPassBegin->renderArea.extent.width >= fbInfo->width &&
+          pRenderPassBegin->renderArea.extent.height >= fbInfo->height && fbInfo->layers == 1;
+
+      for(size_t i = 0; fbInfo->imageAttachments[i].barrier.sType; i++)
       {
-        VkResourceRecord *att = fb->imageAttachments[i].record;
+        VkResourceRecord *att = fbInfo->imageAttachments[i].record;
         if(att == NULL)
           break;
 
-        record->MarkImageViewFrameReferenced(att, ImageRange(), eFrameRef_ReadBeforeWrite);
+        FrameRefType refType = eFrameRef_ReadBeforeWrite;
+
+        if(renderArea_covers_entire_framebuffer)
```

Comment:
(Not sure where better to put this comment).

This check is necessary but not sufficient. We also need to know if the framebuffer covers the whole image - you can set up a framebuffer that writes to one subset of an image while another part of the image is untouched and potentially later read. This is theoretically possible with width/height but would be rather obscure, but is more likely when the framebuffer only covers some subresources like only one layer in a multi-layer framebuffer.

I _think_ we should be able to cache this at framebuffer creation time. The only wrinkle is for multiview where the framebuffer is created with `layers == 1` but the component image views have N layers for N-way multiview. I think we'd need to defer the decision of whether a framebuffer covers the whole attachment until here, because a framebuffer with `layers == 1` and an attachment with 2 layers **is** complete write when used with a renderpass with 2x multiview, but **is not** complete write when used with a renderpass with no multiview (because the second layer is then not written to).

Note we need to check the properties of the underlying image, not just the view (which could itself refer to only a subset of the image).

Reply:
Yeah after looking at how multiview is setup I think I understand how to do this check. I'll update with the new version.

Reply:
I think this is almost there, it's just missing the check that the view covers the whole image. Currently the checks look good for verifying that the RP & FB fully cover the image view attachments, but you could have an image view that only refers to a subset of an image that would still pass and falsely by identified as completely covering.